src/core/sock/sockinfo_tcp.cpp

```diff
@@ -2762,7 +2774,27 @@ int sockinfo_tcp::bind(const sockaddr *__addr, socklen_t __addrlen)
         }
     } else {
         si_tcp_logdbg("OS bind to %s", sockaddr2str(__addr, __addrlen, true).c_str());
-        ret = orig_os_api.bind(m_fd, __addr, __addrlen);
+        if (INPORT_ANY == in_port && m_bind_no_port) {
+            if (m_sock_state == TCP_SOCK_BOUND_NO_PORT) {
+                // bind call from connect()
+                if ((ret = g_bind_no_port->bind_and_set_port_map(m_bound, m_connected, m_fd))) {
+                    UNLOCK_RET(ret);
+                }
+            } else {
+                // first bind call with port 0, we set SO_REUSEPORT so we will be able to bind to a
+                // specific port later when we reuse port
+                int so_reuseport = 1;
+                if ((ret = orig_os_api.setsockopt(m_fd, SOL_SOCKET, SO_REUSEPORT, &so_reuseport,
```

Comment:
Does `SO_REUSEPORT=0` conflict with `IP_BIND_ADDRESS_NO_PORT`

Reply:
no.
SO_REUSEPORT is orthogonal to IP_BIND_ADDRESS_NO_PORT, and set to 0 by default.
but because we are implementing IP_BIND_ADDRESS_NO_PORT without passing it to kernel - we must use SO_REUSEPORT=1 before we call os.bind()